crates/turbo-tasks-memory/src/task.rs

```diff
@@ -174,6 +173,48 @@ struct TaskState {
     last_duration: Duration,
 }
 
+/// Keeps track of emitted and unemitted collectibles. Defaults to None to avoid
+/// allocating memory for two empty hashsets when no collectibles are emitted.
+#[derive(Default)]
+struct MaybeCollectibles {
+    inner: Option<Box<Collectibles>>,
+}
+
+/// The collectibles of a task.
+#[derive(Default)]
+struct Collectibles {
+    emitted: HashSet<(TraitTypeId, RawVc)>,
+    unemitted: HashSet<(TraitTypeId, RawVc)>,
+}
+
+impl MaybeCollectibles {
+    /// Consumes the collectibles (if any) and return them.
+    fn take(&mut self) -> Option<Box<Collectibles>> {
+        self.inner.take()
+    }
+
+    /// Returns a reference to the collectibles (if any).
+    fn as_ref(&mut self) -> Option<&Box<Collectibles>> {
```

Comment:
Nit: Would this allow cleaner code as a `Deref` impl?

Reply:
[The rust documentation for Deref warns:
](https://doc.rust-lang.org/std/ops/trait.Deref.html)
> Deref should only be implemented for smart pointers